icelake/src/catalog/mod.rs

```diff
@@ -189,9 +265,158 @@ pub enum MetadataUpdate {
     },
 }
 
+impl MetadataUpdate {
+    fn apply(&self, metadata: &mut TableMetadata) -> Result<()> {
+        match self {
+            MetadataUpdate::AddSnapshot { snapshot } => metadata.append_snapshot(snapshot.clone()),
+            other => Err(Error::new(
+                ErrorKind::IcebergFeatureUnsupported,
+                format!("update {other} is not supported"),
+            )),
+        }
+    }
+}
+
 /// Update table request.
 pub struct UpdateTable {
     table_name: TableIdentifier,
     requirements: Vec<UpdateRquirement>,
     updates: Vec<MetadataUpdate>,
 }
+
+/// Update table builder.
+pub struct UpdateTableBuilder(UpdateTable);
+
+impl UpdateTable {
+    /// Creates a update table builder.
+    pub fn builder(table_name: TableIdentifier) -> UpdateTableBuilder {
+        UpdateTableBuilder(UpdateTable {
+            table_name,
+            requirements: vec![],
+            updates: vec![],
+        })
+    }
+}
+
+impl UpdateTableBuilder {
+    /// Add requirements.
+    pub fn add_requirements(
+        &mut self,
+        requirements: impl IntoIterator<Item = UpdateRquirement>,
+    ) -> &mut Self {
+        self.0.requirements.extend(requirements);
+        self
+    }
+
+    /// Add updates.
+    pub fn add_updates(&mut self, updates: impl IntoIterator<Item = MetadataUpdate>) -> &mut Self {
+        self.0.updates.extend(updates);
+        self
+    }
+
+    /// Build.
+    pub fn build(self) -> UpdateTable {
+        self.0
+    }
+}
+
+/// Args for creating opendal operator
+pub enum OperatorArgs {
```

Comment:
Operator args could be very large, how about storing as `HashMap<String, String>` and build `Operator` via `Operator::from_map()`?

Reply:
Really cool, I didn't realize this method before.